### PR TITLE
Improve device and code compute capability check

### LINKED ISSUE #164596: [CUDA 13] Incorrect warning messages are printed out upon importing torch on newer devices like Spark
### 🐛 Describe the bug

Example warning:

0: /usr/local/lib/python3.12/dist-packages/torch/cuda/__init__.py:283: UserWarning:  0:     Found GPU0 NVIDIA GB10 which is of cuda capability 12.1. 0:     Minimum and Maximum cuda capability supported by this version of PyTorch is 0:     (8.0) - (12.0)

The above warning is incorrect because pytorch binary compiled with 12.0 can directly run on 12.1 (Spark)

Workaround PR: https://github.com/pytorch/pytorch/pull/164590  

### Versions

Nightly cuda13 binary. 

cc @ptrblck @eqy @tinglvv @atalman @malfet

### LINKED ISSUE #164708: `torch.cuda._check_capability` could raise a false-negative warning
The function in question is defined here -- 
https://github.com/pytorch/pytorch/blob/9fff8155c362da777e7ce31b85fb2dc7cfced2d5/torch/cuda/__init__.py#L247-L283

In my environment, PyTorch is installed with the following env vars: 
- `TORCH_CUDA_ARCH_LIST="8.9 9.0+PTX"`
- `CUDA_HOME=/usr/local/cuda-12.9`
- `CUDNN_LIB_DIR=/usr/local/cuda-12.9/targets/x86_64-linux/lib`
- `CUDNN_INCLUDE_DIR=/usr/local/cuda-12.9/targets/x86_64-linux/include`

Thus my source build PyTorch works in my environment but when I do something on CUDA then I see a warning as follows:
```
$ python -c 'import torch; print(torch.ones((1,), device="cuda"))'
/path/to/torch/cuda/__init__.py:284: UserWarning:
    Found GPU0 NVIDIA RTX 6000 Ada Generation which is of cuda capability 8.9.
    Minimum and Maximum cuda capability supported by this version of PyTorch is
    (8.9) - (9.0)

  warnings.warn(
/path/to/torch/cuda/__init__.py:302: UserWarning:
    Please install PyTorch with a following CUDA
    configurations:  12.6 12.8 13.0 following instructions at
    https://pytorch.org/get-started/locally/

  warnings.warn(matched_cuda_warn.format(matched_arches))
tensor([1.], device='cuda:0')
```

I'm wondering if this check should respect the minor version of the device being checked, not only the major version.

cc @ptrblck @msaroufim @eqy @jerryzh168 @malfet

### PR DESCRIPTION
**Pull request by lakshayg**
Here's what the revised warning looks like

```
torch/cuda/__init__.py:340: UserWarning: Found GPU0 ... which is of compute capability (CC) 9.0.
The following list shows the CCs this version of PyTorch was built for and the hardware CCs it supports:
- 7.5 which supports hardware CC >=7.5,<8.0
- 8.0 which supports hardware CC >=8.0,<9.0 except {8.7}
- 8.6 which supports hardware CC >=8.6,<9.0 except {8.7}
- 10.0 which supports hardware CC >=10.0,<11.0 except {10.1}
- 10.1 which supports hardware CC {10.1, 11.0}
Please follow the instructions at https://pytorch.org/get-started/locally/ to install a PyTorch release that supports one of these CUDA versions: 12.6, 12.8, 13.0
```

fixes: #164708

Reference: NVIDIA Compute Capability to Device Mapping: https://developer.nvidia.com/cuda-gpus 

cc: @ptrblck @nWEIdia @tinglvv

### CODE REVIEW

File: torch/cuda/__init__.py [hunk 1]
```diff
@@ -245,65 +245,113 @@ def _extract_arch_version(arch_string: str) -> int:
...
+    103: CompatInterval(start=103),
+    110: CompatInterval(start=110),
+    120: CompatInterval(start=120),
+}
+
+
+# TORCH_CUDA_ARCH_LIST for PyTorch releases
+PYTORCH_RELEASES_CODE_CC: dict[str, set[int]] = {
+    "12.6": {50, 60, 70, 80, 86, 90},
+    "12.8": {70, 80, 86, 90, 100, 110, 120},
```
Comment (nWEIdia) [inline torch/cuda/__init__.py]: According to current setup: 
https://github.com/pytorch/pytorch/blob/fdcf402d82d32cc5122b4c39dc674ab6f8e211ce/.ci/aarch64_linux/aarch64_ci_build.sh#L9-L14 

110 would only be available in standard build from 13.0 binary. So the warning message, when on THOR, should only prompt users to download cu130 binary, not cu128.

**Comment by nWEIdia** [comment]
> @pytorchbot label "topic: not user facing"

I would argue this is `user facing` , as potentially this code path could produce something the user might not be seeing previously. Or it fixes something user was seeing (e.g. on Spark), which was the main motivation of this PR. The user won't see the warning anymore e.g. while running pytorch nightly cu13.0 binary with this PR in.

File: torch/cuda/__init__.py (hunk 1 above)
Comment (lakshayg) [reply torch/cuda/__init__.py]: I have fixed this. I see that this scripts also says that 12.6 only supports SM>=8.0. Should I update that as well or is there something I'm missing?

**Comment by lakshayg** [comment]
> > @lakshayg: @pytorchbot label "topic: not user facing"
> 
> @nWEIdia: I would argue this is `user facing` , ...

I have added the "release notes: cuda" label. Is that the right one or do you recommend a different one?

**Review by nWEIdia — APPROVED** [review]
LGTM, thanks!

**Review by atalman — APPROVED** [review]
lgtm, do we have a link to Nvidia documentation where device and code compute capability is defined ? This way will be easier to update the code for future devices

**Comment by nWEIdia** [comment]
Yes @atalman : I updated the PR description with this official link: https://developer.nvidia.com/cuda-gpus

**Comment by lakshayg** [comment]
@nWEIdia I believe @ptrblck wanted to review it as well. He wanted to think through the case when someone is building pytorch with custom arch list.

**Comment by atalman** [comment]
Hi @lakshayg is this ready to be landed ?

**Comment by lakshayg** [comment]
@atalman it's ready IMO but @ptrblck might want to take a look

File: test/test_cuda_compatibility.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+    @patch("torch.cuda.get_arch_list", return_value=["sm_60"])
+    @patch("torch.cuda.get_device_capability", return_value=(7, 0))
+    def test_warning_suggests_compatible_pytorch_release(self, *args):
+        with warnings.catch_warnings(record=True) as w:
+            warnings.simplefilter("always")
+            torch.cuda._check_capability()
+            self.assertEqual(len(w), 1)
+            msg = str(w[0].message)
+            self.assertIn("12.6", msg)
```
Comment (ptrblck) [inline test/test_cuda_compatibility.py:139]: Could generalize the tests a it more as they are hard-coding the current support matrix and will break in the future e.g. when 13.0 is replaced with 13.1 etc.?

File: torch/cuda/__init__.py [hunk 3]
```diff
@@ -253,68 +253,136 @@ def _extract_arch_version(arch_string: str) -> int:
...
+    110: _CompatInterval(start=110),
+    120: _CompatInterval(start=120),
+}
+
+
+# TORCH_CUDA_ARCH_LIST for PyTorch releases
+PYTORCH_RELEASES_CODE_CC: dict[str, set[int]] = {
+    "12.6": {50, 60, 70, 80, 86, 90},
+    "12.8": {70, 80, 86, 90, 100, 120},
+    "13.0": {75, 80, 86, 90, 100, 110, 120},
```
Comment (ptrblck) [inline torch/cuda/__init__.py:321]: Do we need to split between x86 and SBSA since `sm_110` is SBSA only?

Comment (lakshayg) [reply torch/cuda/__init__.py:321]: I don't think this part needs any change because it's only used to report the compatible pytorch release. If the user has sm_110, they'll see pytorch cuda 13.0 as the suggested release. This mapping only ever goes from right to left.

File: test/test_cuda_compatibility.py (hunk 2 above)
Comment (lakshayg) [reply test/test_cuda_compatibility.py:139]: I have mocked the relevant dict so the test will not need to be changed

**Comment by busFred** [comment]
i'm on spark with torch 2.9.1 cu13.0 and i'm also receiving the same error. i skim through this discussion chain, correct me if i'm wrong, it seems this warning is just a false alarm and does not affect actual usability; I guess I could just ignore this until the next update comes out(?)

**Comment by ptrblck** [comment]
@busFred Yes, ignore the warning assuming you are already using any of our PyTorch builds with CUDA 13.x on Spark.
